File: backend/services/webhooks-server/src/main/scala/edu/ucdavis/fiehnlab/mona/backend/core/persistence/rest/server/webhooks/types/WebHook.scala
package edu.ucdavis.fiehnlab.mona.backend.core.persistence.rest.server.webhooks.types

import java.util.Date

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document

import scala.annotation.meta.field

/**
  * defines an internal or external webhook to be notified in case of observed events
  *
  * @param name
  */
@Document(collection = "WEBHOOK")
case class WebHook(
                    @(Id@field)
                    name: String,
                    url: String,
                    description: String = "None provided",
                    username: String
                  )

/**
  * the result of a triggered webhook
  *
  * @param url
  * @param success
  * @param error
  * @param invoked when was this hook executed last
  */
@Document(collection = "WEBHOOKS_TRIGGERED")
case class WebHookResult(
                          @(Id@field)
                          id: String,
                          @(Indexed@field)
                          name: String,
                          url: String,
                          success: Boolean = true,
                          error: String = "",
                          @(Indexed@field)
                          invoked: Date = new Date()
                        )

File: backend/services/webhooks-server/src/main/java/edu/ucdavis/fiehnlab/mona/backend/core/persistence/rest/server/webhooks/domain/WebHook.java
package edu.ucdavis.fiehnlab.mona.backend.core.persistence.rest.server.webhooks.domain;


import org.springframework.context.annotation.Profile;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;
import java.io.Serializable;
import java.util.Objects;

@Entity
@Table(name = "webhook")
@Profile({"mona.persistence"})
public class WebHook implements Serializable {
    @Id
    private String name;
    private String url;
    private String description = "None provided";
    private String emailAddress;

    public WebHook(String name, String url, String description, String emailAddress) {
        this.name = name;
        this.url = url;
        this.description = description;
        this.emailAddress = emailAddress;
    }

    public WebHook() {
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getEmailAddress() {
        return emailAddress;
    }

    public void setUsername(String emailAddress) {
        this.emailAddress = emailAddress;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        WebHook webHook = (WebHook) o;
        return Objects.equals(name, webHook.name) && Objects.equals(url, webHook.url) && Objects.equals(description, webHook.description) && Objects.equals(emailAddress, webHook.emailAddress);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, url, description, emailAddress);
    }
}
